File: implementations/typescript/dijkstra.ts
/** Trusted, repository-controlled Dijkstra reference implementation. */
export type NodeId = string;

export interface Edge {
  to: NodeId;
  weight: number;
}

export type Graph = ReadonlyMap<NodeId, readonly Edge[]>;

export interface DijkstraResult {
  found: boolean;
  cost: number | null;
  path: NodeId[];
}

interface QueueEntry {
  node: NodeId;
  distance: number;
}

class MinHeap {
  private readonly entries: QueueEntry[] = [];

  get size(): number {
    return this.entries.length;
  }

  push(entry: QueueEntry): void {
    this.entries.push(entry);
    let child = this.entries.length - 1;
    while (child > 0) {
      const parent = Math.floor((child - 1) / 2);
      if (!this.before(this.entries[child], this.entries[parent])) break;
      [this.entries[parent], this.entries[child]] = [
        this.entries[child],
        this.entries[parent],
      ];
      child = parent;
    }
  }

  pop(): QueueEntry | undefined {
    const first = this.entries[0];
    const last = this.entries.pop();
    if (!first || !last || this.entries.length === 0) return first;

    this.entries[0] = last;
    let parent = 0;
    while (true) {
      const left = parent * 2 + 1;
      const right = left + 1;
      let next = parent;
      if (
        left < this.entries.length &&
        this.before(this.entries[left], this.entries[next])
      ) {
        next = left;
      }
      if (
        right < this.entries.length &&
        this.before(this.entries[right], this.entries[next])
      ) {
        next = right;
      }
      if (next === parent) break;
      [this.entries[parent], this.entries[next]] = [
        this.entries[next],
        this.entries[parent],
      ];
      parent = next;
    }
    return first;
  }

  private before(a: QueueEntry, b: QueueEntry): boolean {
    return (
      a.distance < b.distance || (a.distance === b.distance && a.node < b.node)
    );
  }
}

function validateGraph(graph: Graph, start: NodeId, goal: NodeId): void {
  if (!graph.has(start) || !graph.has(goal)) {
    throw new Error("Start and goal must be present in the graph");
  }
  for (const edges of graph.values()) {
    for (const edge of edges) {
      if (!graph.has(edge.to)) throw new Error(`Unknown node: ${edge.to}`);
      if (!Number.isFinite(edge.weight) || edge.weight < 0) {
        throw new Error("Dijkstra requires finite, nonnegative weights");
      }
    }
  }
}

export function dijkstra(
  graph: Graph,
  start: NodeId,
  goal: NodeId,
): DijkstraResult {
  validateGraph(graph, start, goal);

  const distances = new Map<NodeId, number>();
  const parents = new Map<NodeId, NodeId>();
  for (const node of graph.keys())
    distances.set(node, Number.POSITIVE_INFINITY);
  distances.set(start, 0);

  const frontier = new MinHeap();
  frontier.push({ node: start, distance: 0 });

  while (frontier.size > 0) {
    const current = frontier.pop()!;
    if (current.distance !== distances.get(current.node)) continue;
    if (current.node === goal) break;

    for (const edge of graph.get(current.node) ?? []) {
      const candidate = current.distance + edge.weight;
      if (candidate >= distances.get(edge.to)!) continue;

      distances.set(edge.to, candidate);
      parents.set(edge.to, current.node);
      frontier.push({ node: edge.to, distance: candidate });
    }
  }

  const goalDistance = distances.get(goal)!;
  if (!Number.isFinite(goalDistance)) {
    return { found: false, cost: null, path: [] };
  }

  const path: NodeId[] = [];
  let node: NodeId | undefined = goal;
  while (node !== undefined) {
    path.push(node);
    if (node === start) break;
    node = parents.get(node);
  }
  path.reverse();
  return { found: true, cost: goalDistance, path };
}

File: implementations/java/Dijkstra.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

/** Trusted, repository-controlled Dijkstra reference implementation. */
public final class Dijkstra {
    public record Edge(String to, double weight) {}

    public record Result(boolean found, Double cost, List<String> path) {}

    private record QueueEntry(String node, double distance) {}

    private static void validateGraph(
            Map<String, List<Edge>> graph, String start, String goal) {
        if (!graph.containsKey(start) || !graph.containsKey(goal)) {
            throw new IllegalArgumentException("start and goal must be present in the graph");
        }
        for (List<Edge> edges : graph.values()) {
            for (Edge edge : edges) {
                if (!graph.containsKey(edge.to())) {
                    throw new IllegalArgumentException("unknown node: " + edge.to());
                }
                if (!Double.isFinite(edge.weight()) || edge.weight() < 0) {
                    throw new IllegalArgumentException(
                            "Dijkstra requires finite, nonnegative weights");
                }
            }
        }
    }

    public static Result shortestPath(
            Map<String, List<Edge>> graph, String start, String goal) {
        validateGraph(graph, start, goal);

        Map<String, Double> distances = new HashMap<>();
        Map<String, String> parents = new HashMap<>();
        for (String node : graph.keySet()) {
            distances.put(node, Double.POSITIVE_INFINITY);
        }
        distances.put(start, 0.0);

        PriorityQueue<QueueEntry> frontier = new PriorityQueue<>(
                Comparator.comparingDouble(QueueEntry::distance)
                        .thenComparing(QueueEntry::node));
        frontier.add(new QueueEntry(start, 0.0));

        while (!frontier.isEmpty()) {
            QueueEntry current = frontier.remove();
            if (current.distance() != distances.get(current.node())) {
                continue;
            }
            if (current.node().equals(goal)) {
                break;
            }

            for (Edge edge : graph.get(current.node())) {
                double candidate = current.distance() + edge.weight();
                if (candidate >= distances.get(edge.to())) {
                    continue;
                }

                distances.put(edge.to(), candidate);
                parents.put(edge.to(), current.node());
                frontier.add(new QueueEntry(edge.to(), candidate));
            }
        }

        double goalDistance = distances.get(goal);
        if (Double.isInfinite(goalDistance)) {
            return new Result(false, null, List.of());
        }

        List<String> path = new ArrayList<>();
        for (String node = goal; node != null; node = parents.get(node)) {
            path.add(node);
            if (node.equals(start)) {
                break;
            }
        }
        Collections.reverse(path);
        return new Result(true, goalDistance, List.copyOf(path));
    }

    public static void main(String[] args) {
        Map<String, List<Edge>> graph = Map.of(
                "A", List.of(new Edge("B", 2), new Edge("C", 7)),
                "B", List.of(new Edge("C", 1), new Edge("D", 5)),
                "C", List.of(new Edge("D", 1)),
                "D", List.of());
        Result result = shortestPath(graph, "A", "D");
        if (!result.equals(new Result(true, 4.0, List.of("A", "B", "C", "D")))) {
            throw new AssertionError(result);
        }
        System.out.println("java dijkstra: ok");
    }
}

